File: lib/src/config.test.ts
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { bucketsOverlap, loadConfig, type BucketConfig } from "./config.js";

const ORIGINAL_ENV = process.env;

beforeEach(() => {
  process.env = { ...ORIGINAL_ENV };
});

afterEach(() => {
  process.env = ORIGINAL_ENV;
});

const base: BucketConfig = {
  bucket: "videos",
  endpoint: "https://r2.example.com",
  accessKeyId: "ak",
  secretAccessKey: "sk",
  region: "auto",
};

describe("bucketsOverlap", () => {
  it("different endpoint → not overlap", () => {
    expect(bucketsOverlap(base, { ...base, endpoint: "https://other.example" })).toBe(false);
  });

  it("different bucket name → not overlap", () => {
    expect(bucketsOverlap(base, { ...base, bucket: "other" })).toBe(false);
  });

  it("same bucket, no prefixes → overlap", () => {
    expect(bucketsOverlap(base, base)).toBe(true);
  });

  it("one prefix subsumes the other → overlap (both directions)", () => {
    const a: BucketConfig = { ...base, prefix: "uploads/" };
    const b: BucketConfig = { ...base, prefix: "uploads/transcoded/" };
    expect(bucketsOverlap(a, b)).toBe(true);
    expect(bucketsOverlap(b, a)).toBe(true);
  });

  it("disjoint prefixes → not overlap", () => {
    const a: BucketConfig = { ...base, prefix: "uploads/" };
    const b: BucketConfig = { ...base, prefix: "transcoded/" };
    expect(bucketsOverlap(a, b)).toBe(false);
  });

  it("empty prefix subsumes any other prefix → overlap", () => {
    const a: BucketConfig = { ...base }; // no prefix
    const b: BucketConfig = { ...base, prefix: "anything/" };
    expect(bucketsOverlap(a, b)).toBe(true);
  });

  it("normalizes endpoint case", () => {
    const a: BucketConfig = { ...base, endpoint: "https://R2.Example.Com" };
    const b: BucketConfig = { ...base, endpoint: "https://r2.example.com" };
    expect(bucketsOverlap(a, b)).toBe(true);
  });

  it("normalizes endpoint with trailing path", () => {
    const a: BucketConfig = { ...base, endpoint: "https://r2.example.com/" };
    const b: BucketConfig = { ...base, endpoint: "https://r2.example.com" };
    expect(bucketsOverlap(a, b)).toBe(true);
  });
});

describe("loadConfig destination validation", () => {
  it("rejects multiple pairs that share the same destination bucket namespace", () => {
    process.env.BUCKETS_CONFIG = JSON.stringify([
      {
        source: { bucket: "src-a", endpoint: "https://r2.example.com" },
        dest: { bucket: "shared-dest", endpoint: "https://R2.Example.com/path" },
        accessKeyId: "ak",
        secretAccessKey: "sk",
      },
      {
        source: { bucket: "src-b", endpoint: "https://r2.example.com" },
        dest: { bucket: "shared-dest", endpoint: "https://r2.example.com" },
        accessKeyId: "ak",
        secretAccessKey: "sk",
      },
    ]);

    expect(() => loadConfig("local")).toThrow(/Destination buckets must be unique/);
  });

  it("allows multiple pairs with different destination buckets", () => {
    process.env.BUCKETS_CONFIG = JSON.stringify([
      {
        source: { bucket: "src-a", endpoint: "https://r2.example.com" },
        dest: { bucket: "dest-a", endpoint: "https://r2.example.com" },
        accessKeyId: "ak",
        secretAccessKey: "sk",
      },
      {
        source: { bucket: "src-b", endpoint: "https://r2.example.com" },
        dest: { bucket: "dest-b", endpoint: "https://r2.example.com" },
        accessKeyId: "ak",
        secretAccessKey: "sk",
      },
    ]);

    expect(loadConfig("local").pairs).toHaveLength(2);
  });
});

File: core/config_test.go
package core

import (
	"strings"
	"testing"
)

func TestBucketsOverlap(t *testing.T) {
	base := BucketConfig{Bucket: "videos", Endpoint: "https://r2.example.com", AccessKeyID: "ak", SecretAccessKey: "sk", Region: "auto"}
	with := func(mut func(*BucketConfig)) BucketConfig {
		c := base
		mut(&c)
		return c
	}

	cases := []struct {
		name string
		a, b BucketConfig
		want bool
	}{
		{"different endpoint", base, with(func(c *BucketConfig) { c.Endpoint = "https://other.example" }), false},
		{"different bucket", base, with(func(c *BucketConfig) { c.Bucket = "other" }), false},
		{"same, no prefixes", base, base, true},
		{"prefix subsumes a→b", with(func(c *BucketConfig) { c.Prefix = "uploads/" }), with(func(c *BucketConfig) { c.Prefix = "uploads/transcoded/" }), true},
		{"prefix subsumes b→a", with(func(c *BucketConfig) { c.Prefix = "uploads/transcoded/" }), with(func(c *BucketConfig) { c.Prefix = "uploads/" }), true},
		{"disjoint prefixes", with(func(c *BucketConfig) { c.Prefix = "uploads/" }), with(func(c *BucketConfig) { c.Prefix = "transcoded/" }), false},
		{"empty prefix subsumes", base, with(func(c *BucketConfig) { c.Prefix = "anything/" }), true},
		{"endpoint case-insensitive", with(func(c *BucketConfig) { c.Endpoint = "https://R2.Example.Com" }), base, true},
		{"endpoint trailing slash", with(func(c *BucketConfig) { c.Endpoint = "https://r2.example.com/" }), base, true},
	}
	for _, c := range cases {
		if got := BucketsOverlap(c.a, c.b); got != c.want {
			t.Errorf("%s: BucketsOverlap = %v, want %v", c.name, got, c.want)
		}
	}
}

func TestLoadConfigRejectsSharedDestinations(t *testing.T) {
	t.Setenv("BUCKETS_CONFIG_FILE", "")
	t.Setenv("BUCKETS_CONFIG", `[
		{"source":{"bucket":"src-a","endpoint":"https://r2.example.com"},"dest":{"bucket":"shared-dest","endpoint":"https://R2.Example.com/path"},"accessKeyId":"ak","secretAccessKey":"sk"},
		{"source":{"bucket":"src-b","endpoint":"https://r2.example.com"},"dest":{"bucket":"shared-dest","endpoint":"https://r2.example.com"},"accessKeyId":"ak","secretAccessKey":"sk"}
	]`)
	_, err := LoadConfig(PlatformLocal)
	if err == nil {
		t.Fatal("expected shared destination buckets to be rejected")
	}
	if !strings.Contains(err.Error(), "destination buckets must be unique") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestLoadConfigAllowsDistinctDestinations(t *testing.T) {
	t.Setenv("BUCKETS_CONFIG_FILE", "")
	t.Setenv("BUCKETS_CONFIG", `[
		{"source":{"bucket":"src-a","endpoint":"https://r2.example.com"},"dest":{"bucket":"dest-a","endpoint":"https://r2.example.com"},"accessKeyId":"ak","secretAccessKey":"sk"},
		{"source":{"bucket":"src-b","endpoint":"https://r2.example.com"},"dest":{"bucket":"dest-b","endpoint":"https://r2.example.com"},"accessKeyId":"ak","secretAccessKey":"sk"}
	]`)
	cfg, err := LoadConfig(PlatformLocal)
	if err != nil {
		t.Fatalf("LoadConfig: %v", err)
	}
	if len(cfg.Pairs) != 2 {
		t.Errorf("pairs = %d, want 2", len(cfg.Pairs))
	}
}
